src/graphtypes/simplegraphs/simpleedgeiter.jl

```diff
@@ -5,83 +5,82 @@ The function [`edges`](@ref) returns a `SimpleEdgeIter` for `AbstractSimpleGraph
+        s += one(T)
+    end
+    return SimpleEdgeIterState(zero(T), 1)
+end
 
-function _next(
-    eit::SimpleEdgeIter{T},
-    state::SimpleEdgeIterState{T} = SimpleEdgeIterState(one(T), 1, false),
-    first::Bool = true) where T <: Integer
+function edge_next(g::Union{SimpleGraph{T}, SimpleDiGraph{T}}, 
+    state::SimpleEdgeIterState{T}) where {T <: Integer}
     s = state.s
     di = state.di
-    if !first
-        di += 1
-    end
-    fin = state.fin
-    while s <= length(eit.adj)
-        arr = eit.adj[s]
-        while di <= length(arr)
-            if eit.directed || s <= arr[di]
-                return SimpleEdgeIterState(s, di, fin)
+    e = SimpleEdge(s, fadj(g, s)[di])
+    di += 1
+    while s <= nv(g)
+        sadj = fadj(g, s)
+        while di <= length(sadj)
+            if is_directed(g) || s <= sadj[di]
+                return e, SimpleEdgeIterState(s, di)
             end
             di += 1
         end
         s += one(T)
         di = 1
     end
-    fin = true
-    return SimpleEdgeIterState(s, di, fin)
+    return e, SimpleEdgeIterState(zero(T), 1)
 end
 
-start(eit::SimpleEdgeIter) = _next(eit)
-done(eit::SimpleEdgeIter, state::SimpleEdgeIterState) = state.fin
-length(eit::SimpleEdgeIter) = eit.m
-
-function next(eit::SimpleEdgeIter, state::SimpleEdgeIterState)
-    edge = SimpleEdge(state.s, eit.adj[state.s][state.di])
-    return(edge, _next(eit, state, false))
-end
+start(eit::SimpleEdgeIter) = edge_start(eit.g)
+done(eit::SimpleEdgeIter, state::SimpleEdgeIterState) = state.s == 0
+length(eit::SimpleEdgeIter) = ne(eit.g)
+next(eit::SimpleEdgeIter, state::SimpleEdgeIterState) = edge_next(eit.g, state)
 
 function _isequal(e1::SimpleEdgeIter, e2)
+    k = 0
     for e in e2
-        s, d = Tuple(e)
-        found = length(searchsorted(e1.adj[s], d)) > 0
-        if !e1.directed
-            found = found || length(searchsorted(e1.adj[d], s)) > 0
-        end
-        !found && return false
+        has_edge(e1.g, e) || return false
+        k += 1
     end
-    return true
+    return k == ne(e1.g)
 end
 ==(e1::SimpleEdgeIter, e2::AbstractVector{SimpleEdge}) = _isequal(e1, e2)
 ==(e1::AbstractVector{SimpleEdge}, e2::SimpleEdgeIter) = _isequal(e2, e1)
 ==(e1::SimpleEdgeIter, e2::Set{SimpleEdge}) = _isequal(e1, e2)
 ==(e1::Set{SimpleEdge}, e2::SimpleEdgeIter) = _isequal(e2, e1)
 
-
-function ==(e1::SimpleEdgeIter, e2::SimpleEdgeIter)
-    length(e1.adj) == length(e2.adj) || return false
-    e1.directed == e2.directed || return false
-    for i in 1:length(e1.adj)
-        e1.adj[i] == e2.adj[i] || return false
+function ==(e1::SimpleEdgeIter, e2::SimpleEdgeIter) 
+    g = e1.g
+    h = e2.g
+    g.ne == g.ne || return false
+    m = min(nv(g), nv(h))
+    for i in 1:m
+        fadj(g, i) == fadj(h, i) || return false
+    end
+    nv(g) == nv(h) && return true
+    for i in m+1:nv(g)
+        isempty(fadj(g, i)) || return false
+    end
+    for i in m+1:nv(h)
+        isempty(fadj(h, i)) || return false
```

Comment:
Nice, elegant way of doing this.

Reply:
I don't think refactoring this out should block this PR but it would probably be a good followup if people want to compare graphs like this. I don't think it should be the default (some applications want to distinguish graphs with empty vertices on the end)

Reply:
When two graphs are compared we check `nv` as well, so this doesn't really impact graph equality. I think that equality of edges should be restricted to edges in the edge set, so this makes intuitive sense to me.